### PR TITLE
15/isna element

### LINKED ISSUE #15: Implement isna_element

### PR DESCRIPTION
**Pull request by chaburkland**
Closes #15.

### CODE REVIEW

File: arraykit.c [hunk 1]
```diff
@@ -61,6 +63,8 @@
         fprintf(stderr, #msg);  \
     _AK_DEBUG_END()
 
+# define isnanj(val) (isnan(val.real) || isnan(val.imag))
+
```
Comment (brandtbucher) [inline arraykit.c:67]: I'm not convinced that this deserves to be a macro. It doesn't really save us any space, the name is a bit misleading, and it hides a potentially nasty double-evaluation.

Maybe just replace its four uses below with the expanded version?

```suggestion
```

File: arraykit.c [hunk 2]
```diff
@@ -270,6 +274,65 @@ resolve_dtype_iter(PyObject *Py_UNUSED(m), PyObject *arg)
...
+        return PyBool_FromLong(PyArrayScalar_VAL(arg, Datetime) == NPY_DATETIME_NAT);
+    }
+
+    // NaT - Timedelta
+    if (PyArray_IsScalar(arg, Timedelta)) {
+        return PyBool_FromLong(PyArrayScalar_VAL(arg, Timedelta) == NPY_DATETIME_NAT);
+    }
+
+    // None
+    return PyBool_FromLong(arg == Py_None);
```
Comment (brandtbucher) [inline arraykit.c:333]: To me, this seems like a *very* common case that is *very* fast to check. Maybe we should put this at the very top of the function, like:

```c
    // None
    if (arg == Py_None) {
        Py_RETURN_TRUE;
    }
```

And then just return `False` if we reach the bottom:

```suggestion
    Py_RETURN_FALSE;
```

**Review by brandtbucher — COMMENTED** [review]
Looks mostly good! You know I'm not a fan of dtype branching, but 15x speedup is 15x speedup. Haha.

(I just reviewed the C code for now. I'll get to the Python code later.)

File: arraykit.c [hunk 3]
```diff
@@ -270,6 +272,69 @@ resolve_dtype_iter(PyObject *Py_UNUSED(m), PyObject *arg)
...
+    if (PyArray_IsScalar(arg, Complex128)) {
+        npy_cdouble val = PyArrayScalar_VAL(arg, Complex128);
+        return PyBool_FromLong((isnan(val.real) || isnan(val.imag)));
+    }
+    # ifdef PyComplex256ArrType_Type
+    if (PyArray_IsScalar(arg, Complex256)) {
+        npy_clongdouble val = PyArrayScalar_VAL(arg, Complex256);
+        return PyBool_FromLong((isnan(val.real) || isnan(val.imag)));
+    }
+    # endif
```
Comment (brandtbucher) [inline arraykit.c:323]: You've got extra parens.

```suggestion
    if (PyComplex_Check(arg)) {
        Py_complex val = ((PyComplexObject*)arg)->cval;
        return PyBool_FromLong(isnan(val.real) || isnan(val.imag));
    }
    if (PyArray_IsScalar(arg, Complex64)) {
        npy_cfloat val = PyArrayScalar_VAL(arg, Complex64);
        return PyBool_FromLong(isnan(val.real) || isnan(val.imag));
    }
    if (PyArray_IsScalar(arg, Complex128)) {
        npy_cdouble val = PyArrayScalar_VAL(arg, Complex128);
        return PyBool_FromLong(isnan(val.real) || isnan(val.imag));
    }
    # ifdef PyComplex256ArrType_Type
    if (PyArray_IsScalar(arg, Complex256)) {
        npy_clongdouble val = PyArrayScalar_VAL(arg, Complex256);
        return PyBool_FromLong(isnan(val.real) || isnan(val.imag));
    }
    # endif
```

File: performance/main.py [hunk 4]
```diff
@@ -250,6 +250,74 @@ class ArrayGOPerfREF(ArrayGOPerf):
...
+        nan = np.nan
+        nanjs = [
+                complex(nan, 0),
+                -complex(nan, 0),
+                complex(-nan, 0),
+                -complex(-nan, 0),
+                complex(0, nan),
+                -complex(0, nan),
+                complex(0, -nan),
+                -complex(0, -nan),
```
Comment (brandtbucher) [inline performance/main.py:277]: These aren't really giving you any more variety than you already have (although you may consider replacing them with cases where both the real and imaginary parts are signed NaNs):

```suggestion
                complex(nan, 0),
                complex(-nan, 0),
                complex(0, nan),
                complex(0, -nan),
```

File: performance/main.py [hunk 5]
```diff
@@ -250,6 +250,74 @@ class ArrayGOPerfREF(ArrayGOPerf):
...
+        self.values = [
+                # Na-elements
+                np.datetime64('NaT'), np.timedelta64('NaT'), None, float('NaN'), -float('NaN'),
+
+                # Non-float, Non-na elements
+                1, 'str', np.datetime64('2020-12-31'), datetime.date(2020, 12, 31), False,
+        ]
+
+        nan = np.nan
+        nanjs = [
```
Comment (brandtbucher) [inline performance/main.py:269]: Maybe `complex_nans`? `nanjs` sort of implies that only the complex part is NaN:

```suggestion
        complex_nans = [
```

File: performance/main.py [hunk 6]
```diff
@@ -250,6 +250,74 @@ class ArrayGOPerfREF(ArrayGOPerf):
...
+            cfloat_classes.append(np.complex256)
+
+        # Append all the different types of nans across dtypes
+        for ctor in float_classes:
+            self.values.append(ctor(nan))
+            self.values.append(ctor(-nan))
+
+        for ctor in cfloat_classes:
+            for nanj in nanjs:
+                self.values.append(ctor(nanj))
```
Comment (brandtbucher) [inline performance/main.py:295]: Here too:

```suggestion
            for complex_nan in complex_nans:
                self.values.append(ctor(complex_nan))
```

File: test/test_util.py [hunk 7]
```diff
@@ -224,9 +227,64 @@ def test_array_deepcopy_c2(self) -> None:
...
+        self.assertTrue(isna_element(FloatSubclass(-nan)))
+
+    def test_isna_element_false(self) -> None:
+        # Test a wide range of float values, with different precision, across types
+        for val in (
+                1e-1000, 1e-309, 1e-39, 1e-16, 1e-5, 0.1, 0., 1.0, 1e5, 1e16, 1e39, 1e309, 1e1000,
+            ):
+            for sign in (1, -1):
+                for ctor in (np.float16, np.float32, np.float64, float):
+                    self.assertFalse(isna_element((ctor(sign * val))))
```
Comment (brandtbucher) [inline test/test_util.py:277]:
```suggestion
                    self.assertFalse(isna_element(ctor(sign * val)))
```

File: test/test_util.py [hunk 8]
```diff
@@ -224,9 +227,64 @@ def test_array_deepcopy_c2(self) -> None:
...
+
+                if hasattr(np, 'float128'):
+                    self.assertFalse(isna_element((np.float128(sign * val))))
```
Comment (brandtbucher) [inline test/test_util.py:280]:
```suggestion
                    self.assertFalse(isna_element(np.float128(sign * val)))
```

File: test/test_util.py [hunk 9]
```diff
@@ -224,9 +227,64 @@ def test_array_deepcopy_c2(self) -> None:
...
+        nanjs = [
+                complex(nan, 0),
+                -complex(nan, 0),
+                complex(-nan, 0),
+                -complex(-nan, 0),
+                complex(0, nan),
+                -complex(0, nan),
+                complex(0, -nan),
+                -complex(0, -nan),
+        ]
```
Comment (brandtbucher) [inline test/test_util.py:244]: Ditto:

```suggestion
        complex_nans = [
                complex(nan, 0),
                complex(-nan, 0),
                complex(0, nan),
                complex(0, -nan),
        ]
```

File: test/test_util.py [hunk 10]
```diff
@@ -224,9 +227,64 @@ def test_array_deepcopy_c2(self) -> None:
...
+        if hasattr(np, 'complex256'):
+            cfloat_classes.append(np.complex256)
+
+        for float_class in float_classes:
+            self.assertTrue(isna_element(float_class(nan)))
+            self.assertTrue(isna_element(float_class(-nan)))
+
+        for cfloat_class in cfloat_classes:
+            for nanj in nanjs:
+                self.assertTrue(isna_element(cfloat_class(nanj)))
```
Comment (brandtbucher) [inline test/test_util.py:260]: Ditto:

```suggestion
            for complex_nan in complex_nans:
                self.assertTrue(isna_element(cfloat_class(complex_nan)))
```

File: test/test_util.py [hunk 11]
```diff
@@ -224,9 +227,64 @@ def test_array_deepcopy_c2(self) -> None:
...
+
+        self.assertTrue(isna_element(float('NaN')))
+        self.assertTrue(isna_element(-float('NaN')))
+        self.assertTrue(isna_element(None))
+
+        class FloatSubclass(float): pass
+        self.assertTrue(isna_element(FloatSubclass(nan)))
+        self.assertTrue(isna_element(FloatSubclass(-nan)))
+
+    def test_isna_element_false(self) -> None:
```
Comment (brandtbucher) [inline test/test_util.py:270]: Did you want to test complex NaNs here too (like you do in the perf tests)?

File: performance/main.py [hunk 12]
```diff
@@ -250,6 +250,74 @@ class ArrayGOPerfREF(ArrayGOPerf):
...
+        for val in (
+                1e-1000, 1e-309, 1e-39, 1e-16, 1e-5, 0.1, 0., 1.0, 1e5, 1e16, 1e39, 1e309, 1e1000,
+            ):
+            for sign in (1, -1):
+                for ctor in float_classes:
+                    self.values.append(ctor(sign * val))
+
+                for ctor in cfloat_classes:
+                    self.values.append(ctor(complex(sign * val, val)))
+                    self.values.append(ctor(complex(val, sign * val)))
```
Comment (brandtbucher) [inline performance/main.py:307]: Maybe something like this is better? Your current code appends duplicate complex values when `sign` is `1`.

```suggestion
            for ctor in float_classes:
                self.values.append(ctor(val))
                self.values.append(ctor(-val))

            for ctor in cfloat_classes:
                self.values.append(ctor(complex(val, val)))
                self.values.append(ctor(complex(-val, val)))
                self.values.append(ctor(complex(val, -val)))
                self.values.append(ctor(complex(-val, -val)))
```

**Review by brandtbucher — APPROVED** [review]
The C code looks good. I have some suggestions for the tests, but they're quite nitpicky, so feel free to ignore them.

Since this changes `setup.py` and our build process a bit, let's wait to merge this until #45 is figured out.

File: performance/main.py (hunk 4 above)
Comment (chaburkland) [reply performance/main.py:277]: Thanks - I was trying to overly thorough here just to be extra confident I caught all different possible nanjs.

**Pull request merged**